File: otUI/SettingsDialog.cs
using System;
using System.Collections.Generic;
using System.IO;
using Gtk;
using otlib;
using UI = Gtk.Builder.ObjectAttribute;

public class SettingsDialog : Gtk.Dialog
{
    Builder builder;
    private AppSettings appSettings;
    [UI] Gtk.RadioButton codeUpperLowerNum;
    [UI] Gtk.RadioButton codeUpperNum;
    [UI] Gtk.RadioButton codeNumerical;
    [UI] Gtk.RadioButton codeCustomBtn;
    [UI] Gtk.RadioButton codeEmoji;

    [UI] Gtk.Entry codeCustom = new Gtk.Entry();

    [UI] Gtk.RadioButton textUpperLowerNum;
    [UI] Gtk.RadioButton textUpperNum;
    [UI] Gtk.RadioButton textNumerical;
    [UI] Gtk.RadioButton textUpperLowerNumSPCPunc;

    [UI] Gtk.RadioButton textCustomBtn;

    [UI] Gtk.Entry textCustom = new Gtk.Entry();

    [UI] Gtk.ComboBoxText RNGDeviceComboBox = new Gtk.ComboBoxText();

    [UI] Gtk.ComboBoxText themeComboBox = new Gtk.ComboBoxText();


    [UI] Gtk.Button btn_help = new Gtk.Button();
    [UI] Gtk.Switch hasPrettyPrint;
    [UI] Gtk.Switch hasPadding;

    public bool stateHasPrettyPrint = true;
    public bool stateHasPadding = true;

    public static SettingsDialog Create(AppSettings appSettings)
    {
        Builder builder = new Builder(null, "otUI.interfaces.SettingsDialog.glade", null);
        return new SettingsDialog(builder, builder.GetObject("settingsdialog").Handle, appSettings);
    }

    protected SettingsDialog(Builder builder, IntPtr handle, AppSettings appSettings) : base(handle)
    {
        this.builder = builder;
        this.appSettings = appSettings;

        builder.Autoconnect(this);
        AddButton("Close", ResponseType.Close);

        if (OperatingSystem.IsWindows())
        {
            RNGDeviceComboBox.Sensitive = false;
        }
        else
        {
            IEnumerable<string> l = Directory.EnumerateFiles("/dev", "*", SearchOption.AllDirectories);

            try
            {
                foreach (string item in l)
                {
                    RNGDeviceComboBox.AppendText(item);
                }
            }
            catch (UnauthorizedAccessException) { }
        }
        LoadConfig();
    }
    private void LoadConfig()
    {

        hasPrettyPrint.Active = appSettings.HasPrettyPrint;
        hasPadding.Active = appSettings.HasPadding;

        switch (appSettings.CodeCharSet)
        {
            case CharSetTypes.EMOJI:
                codeEmoji.Click();
                break;
            case CharSetTypes.UPPER_LOWER_NUMERIC:
                codeUpperLowerNum.Click();
                break;
            case CharSetTypes.UPPER_NUMERIC:
                codeUpperNum.Click();
                break;
            case CharSetTypes.NUMERIC:
                codeNumerical.Click();
                break;
            default:
                codeCustomBtn.Click();
                break;
        }
        switch (appSettings.TextCharSet)
        {
            case CharSetTypes.UPPER_LOWER_NUMERIC_PUNC_SPC:
                textUpperLowerNumSPCPunc.Click();
                break;
            case CharSetTypes.UPPER_LOWER_NUMERIC:
                textUpperLowerNum.Click();
                break;
            case CharSetTypes.UPPER_NUMERIC:
                textUpperNum.Click();
                break;
            case CharSetTypes.NUMERIC:
                textNumerical.Click();
                break;
            default:
                textCustomBtn.Click();
                break;
        }

        if (appSettings.Theme == "Dark")
        {
            themeComboBox.ActiveId = "2";
        }
        else if (appSettings.Theme == "Light")
        {
            themeComboBox.ActiveId = "1";
        }

        codeCustom.Text = appSettings.CodeCharSetCustom;
        textCustom.Text = appSettings.TextCharSetCustom;
    }
    protected void On_Save_clicked(object sender, EventArgs e)
    {
        if (codeCustomBtn.Active)
        {
            appSettings.CodeCharSetCustom = codeCustom.Text;
        }

        if (textCustomBtn.Active)
        {
            appSettings.TextCharSetCustom = textCustom.Text;
        }

        if (themeComboBox.ActiveText == "System")
        {
            appSettings.Theme = null;
        }
        else if (themeComboBox.ActiveText == "Dark")
        {
            appSettings.Theme = "Dark";
        }
        else if (themeComboBox.ActiveText == "Light")
        {
            appSettings.Theme = "Light";
        }

        appSettings.HasPrettyPrint = hasPrettyPrint.Active;
        appSettings.HasPadding = hasPadding.Active;

        appSettings.Write();
        Destroy();
    }

    // Code charset events
    protected void On_codeCustomBtn_clicked(object sender, EventArgs e)
    {
        appSettings.CodeCharSet = CharSetTypes.CUSTOM;
        codeCustom.IsEditable = true;
    }

    protected void On_codeUpperLowerNum_clicked(object sender, EventArgs e)
    {
        appSettings.CodeCharSet = CharSetTypes.UPPER_LOWER_NUMERIC;
        codeCustom.IsEditable = false;
    }

    protected void On_codeUpperNum_clicked(object sender, EventArgs e)
    {
        appSettings.CodeCharSet = CharSetTypes.UPPER_NUMERIC;
        codeCustom.IsEditable = false;
    }

    protected void On_codeNumerical_clicked(object sender, EventArgs e)
    {
        appSettings.CodeCharSet = CharSetTypes.NUMERIC;
        codeCustom.IsEditable = false;
    }

    protected void On_codeEmoji_clicked(object sender, EventArgs e)
    {
        appSettings.CodeCharSet = CharSetTypes.EMOJI;
        textCustom.IsEditable = false;
    }

    // Text charset events

    protected void On_textCustomBtn_clicked(object sender, EventArgs e)
    {
        appSettings.TextCharSet = CharSetTypes.CUSTOM;
        textCustom.IsEditable = true;
    }

    protected void On_textUpperLowerNum_clicked(object sender, EventArgs e)
    {
        appSettings.TextCharSet = CharSetTypes.UPPER_LOWER_NUMERIC;
        textCustom.IsEditable = false;
    }

    protected void On_textUpperNum_clicked(object sender, EventArgs e)
    {
        appSettings.TextCharSet = CharSetTypes.UPPER_NUMERIC;
        textCustom.IsEditable = false;
    }

    protected void On_textNumerical_clicked(object sender, EventArgs e)
    {
        appSettings.TextCharSet = CharSetTypes.NUMERIC;
        textCustom.IsEditable = false;
    }

    protected void On_textUpperLowerNumSPCPunc_clicked(object sender, EventArgs e)
    {
        appSettings.TextCharSet = CharSetTypes.UPPER_LOWER_NUMERIC_PUNC_SPC;
        textCustom.IsEditable = false;
    }






    protected void On_RNGDeviceComboBox_changed(object sender, EventArgs e)
    {
        if (RNGDeviceComboBox.ActiveText == "None")
        {
            appSettings.RngDevicePath = null;
        }
        else
        {
            appSettings.RngDevicePath = RNGDeviceComboBox.ActiveText;
        }
    }

    protected void On_btn_help_clicked(object sender, EventArgs e)
    {
        HelpDialog hd = HelpDialog.Create(otUI.HelpConst.SettingsHelp);
        hd.Run();
        hd.Destroy();
    }



}

File: UI/SettingsDialog.py
import gi

from Core.Constants.Help import SETTINGS
from Core.Constants import Config
from Core.Settings import CharsetTypes, write
from UI import AboutDialog, HelpDialog
from UI.ThemeLoader import load_theme

import os

gi.require_version('Gtk', '3.0')
gi.require_version('Gio', '2.0')
gi.require_version('Pango', '1.0')

from gi.repository import Gio
from gi.repository import Gtk
from gi.repository import Pango

from main import app_settings, resource
from Core import RNGDevice

Gio.Resource._register(resource)


@Gtk.Template(resource_path="/org/onetimepadui/UI/interfaces/SettingsDialog.ui")
class SettingsDialog(Gtk.Dialog):
    __gtype_name__ = "SettingsDialog"

    text_upper_lower_num_spc_punc = Gtk.Template.Child()
    text_upper_lower_num = Gtk.Template.Child()
    text_upper_num = Gtk.Template.Child()
    text_numerical = Gtk.Template.Child()
    text_custom_btn = Gtk.Template.Child()
    text_custom = Gtk.Template.Child()

    code_upper_lower_num = Gtk.Template.Child()
    code_upper_num = Gtk.Template.Child()
    code_numerical = Gtk.Template.Child()
    code_emoji = Gtk.Template.Child()
    code_custom_btn = Gtk.Template.Child()
    code_custom = Gtk.Template.Child()

    rng_device_combo_box = Gtk.Template.Child()
    has_padding = Gtk.Template.Child()
    theme_combo_box = Gtk.Template.Child()
    has_pretty_print = Gtk.Template.Child()

    def __init__(self):
        super().__init__()
        self.init_template()

        if os.name == "posix":
            dev_list = RNGDevice.list()
            for dev_index in range(1, len(dev_list)):
                self.rng_device_combo_box.append(str(dev_index), str(dev_list[dev_index]))
        else:
            self.rng_device_combo_box.props.button_sensitivity = Gtk.SensitivityType.OFF

        self.load_config()

    def load_config(self):
        self.has_pretty_print.do_state_set(self.has_pretty_print, app_settings.has_pretty_print)
        self.has_padding.do_state_set(self.has_padding, app_settings.has_padding)

        self.code_custom.set_text(",".join(app_settings.code_charset_custom))
        self.text_custom.set_text(",".join(app_settings.text_charset_custom))

        match app_settings.text_charset:
            case CharsetTypes.UPPER_LOWER_NUMERIC_PUNC_SPC:
                self.text_upper_lower_num_spc_punc.do_clicked(self.text_upper_lower_num_spc_punc)
            case CharsetTypes.UPPER_LOWER_NUMERIC:
                self.text_upper_lower_num.do_clicked(self.text_upper_lower_num)
            case CharsetTypes.UPPER_NUMERIC:
                self.text_upper_num.do_clicked(self.text_upper_num)
            case CharsetTypes.NUMERIC:
                self.text_numerical.do_clicked(self.text_numerical)
            case _:
                self.text_custom_btn.do_clicked(self.text_custom_btn)
                self.text_custom_btn.do_clicked(self.text_custom_btn)

        match app_settings.code_charset:
            case CharsetTypes.EMOJI:
                self.code_emoji.do_clicked(self.code_emoji)
            case CharsetTypes.UPPER_LOWER_NUMERIC:
                self.code_upper_lower_num.do_clicked(self.code_upper_lower_num)
            case CharsetTypes.UPPER_NUMERIC:
                self.code_upper_num.do_clicked(self.code_upper_num)
            case CharsetTypes.NUMERIC:
                self.code_numerical.do_clicked(self.code_numerical)
            case _:
                self.code_custom_btn.do_clicked(self.code_custom_btn)
                self.code_custom_btn.do_clicked(self.code_custom_btn)

        if app_settings.theme_id == 2:
            self.theme_combo_box.set_active_id("2")
        elif app_settings.theme_id == 1:
            self.theme_combo_box.set_active_id("1")

        self.rng_device_combo_box.set_active_id(str(app_settings.rng_device_id))
        self.theme_combo_box.set_active_id(str(app_settings.theme_id))
        load_theme()

    @Gtk.Template.Callback()
    def onAboutClicked(self, button):
        AboutDialog.main(self)

    @Gtk.Template.Callback()
    def onSaveClicked(self, button):
        write(str(Config.path()), app_settings)
        self.destroy()

    @Gtk.Template.Callback()
    def onHelpClicked(self, button):
        HelpDialog.main(self, SETTINGS)

    @Gtk.Template.Callback()
    def onRNGDeviceComboboxChanged(self, combobox):
        app_settings.rng_device_path = combobox.get_active_text()
        app_settings.rng_device_id = combobox.get_active_id()

    @Gtk.Template.Callback()
    def onTextUpperLowerNumSPCPuncClicked(self, button):
        app_settings.text_charset = CharsetTypes.UPPER_LOWER_NUMERIC_PUNC_SPC
        self.text_custom.set_editable(False)

    @Gtk.Template.Callback()
    def onTextUpperLowerNumClicked(self, button):
        app_settings.text_charset = CharsetTypes.UPPER_LOWER_NUMERIC
        self.text_custom.set_editable(False)

    @Gtk.Template.Callback()
    def onTextUpperNumClicked(self, button):
        app_settings.text_charset = CharsetTypes.UPPER_NUMERIC
        self.text_custom.set_editable(False)

    @Gtk.Template.Callback()
    def onTextNumericalClicked(self, button):
        app_settings.text_charset = CharsetTypes.NUMERIC
        self.text_custom.set_editable(False)

    @Gtk.Template.Callback()
    def onTextCustomBtnClicked(self, button):
        app_settings.text_charset = CharsetTypes.CUSTOM
        self.text_custom.set_editable(True)

    @Gtk.Template.Callback()
    def onCodeUpperLowerNumClicked(self, button):
        app_settings.code_charset = CharsetTypes.UPPER_LOWER_NUMERIC
        self.code_custom.set_editable(False)

    @Gtk.Template.Callback()
    def onCodeUpperNumClicked(self, button):
        app_settings.code_charset = CharsetTypes.UPPER_NUMERIC
        self.code_custom.set_editable(False)

    @Gtk.Template.Callback()
    def onCodeNumericalClicked(self, button):
        app_settings.code_charset = CharsetTypes.NUMERIC
        self.code_custom.set_editable(False)

    @Gtk.Template.Callback()
    def onCodeEmojiClicked(self, button):
        app_settings.code_charset = CharsetTypes.EMOJI
        self.code_custom.set_editable(False)

    @Gtk.Template.Callback()
    def onCodeCustomBtnClicked(self, button):
        app_settings.code_charset = CharsetTypes.CUSTOM
        self.code_custom.set_editable(True)

    @Gtk.Template.Callback()
    def onHasPaddingStateSet(self, switch, state):
        app_settings.has_padding = state

    @Gtk.Template.Callback()
    def onHasPrettyPrintStateSet(self, switch, state):
        app_settings.has_pretty_print = state

    @Gtk.Template.Callback()
    def onTextCustomChanged(self, editable):
        app_settings.text_charset_custom = self.text_custom.get_text().split(",")

    @Gtk.Template.Callback()
    def onCodeCustomChanged(self, editable):
        app_settings.code_charset_custom = self.code_custom.get_text().split(",")

    @Gtk.Template.Callback()
    def OnThemeComboBoxChanged(self, combo_box):
        app_settings.theme_id = combo_box.get_active_id()
        load_theme()

    @Gtk.Template.Callback()
    def onSetFont(self, font_button):
        font = font_button.get_font()
        print(font)




def main():
    dialog = SettingsDialog()
    dialog.show()

    Gtk.main()
